### DISCUSSION TITLE
Sql injection and setUseLiteral(true)

### DISCUSSION THREAD

**Discussion by lpandzic** [querydsl/querydsl#3558] [Q&A] [answered]
Hello, I wanted to check if use of setUseLiteral(true) can in any form cause sql injection?
Based on my testing it shouldn't but I wanted to verify - [javadoc](https://github.com/querydsl/querydsl/blob/cd608f04ed7534b1b62a78a28beb141071d149b6/querydsl-sql/src/main/java/com/querydsl/sql/Configuration.java#L537-L544) is unclear on the matter.

**Comment by jwgmeligmeyling** [comment, marked as answer]
Literals are serialized in the SQLSerializer via `Configuration.asLiteral`, which in turn delegates to `Templates.serialize`, which internally uses `escapeString`. So some effort is made in sanitising the String literal. However,  `StringEscapeUtils.escapeSql` [was removed from `commons.lang` for a reason](https://security.elarlang.eu/cve-2016-10007-and-cve-2016-10008-2-sql-injection-vulnerabilities-in-dotcms-blacklist-defence-bypass.html) and we can expect malicious users can always come up with inputs for very particular database dialects that are difficult to predict. I wouldn't want to take that risk personally. The overhead of parameters is so minimal, I would use them for user input without exception. For in-code constants the potential sql injection is not a risk, but for those I'd use template expressions so you don't have to enable literal rendering at a global level.

You can also validate that literals are only used with primitive values or enums to guard against injection, if your use case is anything but a string.
